harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -28,56 +27,198 @@
+        elif unit == core.Unit.RECORDS:
+            return Record(length)
+        elif unit == core.Unit.BATCHES:
+            return Batch(length)
+        else:
+            raise ValueError(f"unrecognized searcher unit {unit}")
+
+    @staticmethod
+    def _from_values(
+        batches: Optional[int] = None, records: Optional[int] = None, epochs: Optional[int] = None
+    ):
+        if sum((batches is not None, records is not None, epochs is not None)) != 1:
+            raise ValueError(f"invalid length: batches={batches} records={records} epochs={epochs}")
+        if batches:
+            return Batch(batches)
+        elif records:
+            return Record(records)
+        elif epochs:
+            return Epoch(epochs)
+
+    def _divides(self, steps: int):
+        assert self.value > 0, "TrainUnit value must be > 0"
+        if steps == 0:
+            return False
+        return steps % self.value == 0
+
+
+class Epoch(TrainUnit):
+    pass
+
+
+class Batch(TrainUnit):
+    pass
+
+
+class Record(TrainUnit):
+    pass
+
+
+class _TrainStepType(enum.Enum):
+    CHECKPOINT = "CHECKPOINT"
+    REPORT = "REPORT"
+    VALIDATE = "VALIDATE"
+    TRAIN = "TRAIN"
+
+
+class _TrainStep:
+    def __init__(self, step_type: _TrainStepType, unit: TrainUnit):
+        self.step_type = step_type
+        self.unit = unit
+        self.limit_reached = False
+
+
+class ShouldExit(Exception):
+    """
+    ShouldExit breaks out of the top-level train loop from inside function calls.
+    """
+
+    def __init__(self, skip_exit_checkpoint: bool = False):
+        self.skip_exit_checkpoint = skip_exit_checkpoint
+
+
+class _TrialState:
+    def __init__(
+        self,
+        trial_id: int = 0,
+        last_ckpt: int = 0,
+        steps_completed: int = 0,
+        step_id: int = 0,
+        last_val: int = 0,
+        batches_trained: int = 0,
+        epochs_trained: int = 0,
+    ) -> None:
+        # Store TrialID to distinguish between e.g. pause/restart and continue training.
+        self.trial_id = trial_id
+        self.last_ckpt = last_ckpt
+        self.steps_completed = steps_completed
+        self.step_id = step_id
+        self.last_val = last_val
+        self.batches_trained = batches_trained
+        self.epochs_trained = epochs_trained
+
+
+class PyTorchTrialController:
+    def __init__(
+        self,
+        trial_inst: det.Trial,
+        context: pytorch.PyTorchTrialContext,
+        min_checkpoint_period: TrainUnit,
+        min_validation_period: TrainUnit,
+        average_training_metrics: Optional[bool] = True,
+        smaller_is_better: Optional[bool] = True,
+        max_length: Optional[TrainUnit] = None,
+        steps_completed: Optional[int] = 0,
+        latest_checkpoint: Optional[str] = None,
+        local_training: Optional[bool] = False,
+        test_mode: Optional[bool] = False,
+        # XXX: should default be maxsize or 100?
+        scheduling_unit: Optional[int] = sys.maxsize,
```

Comment:
less important, since this is an internal value now...

Calling this "scheduling_unit" doesn't really make any sense anymore, because it's not about the granularity of scheduling in the master.  It's about the maximum size of the training window before reporting metrics and checking for preemption.

Reply:
agreed scheduling unit isn't a great name. i think internally i call it reporting_period somewhere. what do you suggest?

Reply:
train_workload_max_batches?  I dunno

Reply:
Ah, `scheduling_unit` is a field that the experiment config shouldn't have.

So this _does_ need to be configurable in `.fit()`.  And please don't name it `scheduling_unit`.

Reply:
named reporting_period.

Reply:
`reporting_period` is not bad... but it only applies to training metrics.  Is there a name that communicates that?

(technically it applies to preemption today too, but we could actually make preemption apply per-batch, and just cut those training windows short)

Reply:
i think reporting is realistically what it will be useful for. i'm not sure if there's a word that combines reporting and preemption... "workload_period"